src/main/scala/com/wavesplatform/matcher/model/OrderValidator.scala

```diff
@@ -133,17 +144,68 @@ object OrderValidator {
     )
   }
 
+  private[matcher] def getValidFeeAsset(order: Order, assetType: AssetType): Option[AssetId] = {
+    assetType match {
+      case AssetType.AMOUNT    => order.assetPair.amountAsset
+      case AssetType.PRICE     => order.assetPair.priceAsset
+      case AssetType.RECEIVING => order.getReceiveAssetId
+      case AssetType.SPENDING  => order.getSpendAssetId
+    }
+  }
+
+  private[matcher] def getMinValidFee(order: Order, percentSettings: PercentSettings): Long = {
+
+    lazy val receiveAmount = order.getReceiveAmount(order.amount, order.price).explicitGet()
+    lazy val spentAmount   = order.getSpendAmount(order.amount, order.price).explicitGet()
+
+    val amountFactor = percentSettings.assetType match {
+      case AssetType.AMOUNT    => order.amount
+      case AssetType.PRICE     => if (order.orderType == OrderType.BUY) spentAmount else receiveAmount
+      case AssetType.RECEIVING => receiveAmount
+      case AssetType.SPENDING  => spentAmount
+    }
+
+    multiplyLongByDouble(amountFactor, percentSettings.minFee / 100)
```

Comment:
Could you inline the implementation of `multiplyLongByDouble` here? `multiplyLongByDouble` rounds the result using half up. This conflicts with the usual way we multiple longs with doubles: `((x: Long) * (y: Double)).toLong`. `multiplyLongByDouble` has the right behaviour only in `getMinValidFee`.

Reply:
Done